File: go/cdk8s-plus-http-echo/main.go
package main

import (
	"github.com/aws/constructs-go/constructs/v10"
	"github.com/aws/jsii-runtime-go"
	"github.com/cdk8s-team/cdk8s-core-go/cdk8s/v2"
	"github.com/cdk8s-team/cdk8s-plus-go/cdk8splus26/v2"
)

func NewHttpEcho(scope constructs.Construct, id string) cdk8s.Chart {

	chart := cdk8s.NewChart(scope, jsii.String(id), nil)

	ingress := cdk8splus26.NewIngress(chart, jsii.String("ingress"), nil)

	// TODO check why it is required to pass cdk8splus26.HttpIngressPathType_PREFIX
	ingress.AddRule(jsii.String("/"), EchoBackend(chart, "root"), cdk8splus26.HttpIngressPathType_PREFIX)
	ingress.AddRule(jsii.String("/foo"), EchoBackend(chart, "foo"), cdk8splus26.HttpIngressPathType_PREFIX)
	ingress.AddRule(jsii.String("/foo/bar"), EchoBackend(chart, "foo-bar"), cdk8splus26.HttpIngressPathType_PREFIX)

	ingress.AddHostDefaultBackend(jsii.String("my.host"), EchoBackend(chart, "my.host/hey"))

	return chart
}

func EchoBackend(scope constructs.Construct, text string) cdk8splus26.IngressBackend {

	deploy := cdk8splus26.NewDeployment(scope, jsii.String(text), &cdk8splus26.DeploymentProps{
		Containers: &[]*cdk8splus26.ContainerProps{
			{
				Image:      jsii.String("hashicorp/http-echo"),
				Args:       &[]*string{jsii.String("-text"), jsii.String(text)},
				PortNumber: jsii.Number(5678),
			},
		},
	})

	return cdk8splus26.IngressBackend_FromService(deploy.ExposeViaService(&cdk8splus26.DeploymentExposeViaServiceOptions{
		Ports: &[]*cdk8splus26.ServicePort{{Port: jsii.Number(5678)}}}), nil)

}

func main() {
	app := cdk8s.NewApp(nil)
	NewHttpEcho(app, "http-echo")
	app.Synth()
}

File: python/cdk8s-plus-http-echo/main.py
#!/usr/bin/env python3
from constructs import Construct
from cdk8s import App, Chart

import cdk8s_plus_26 as kplus


class HttpEcho(Chart):
    def __init__(self, scope: Construct, id: str):
        super().__init__(scope, id)

        ingress = kplus.Ingress(self, "ingress")
        ingress.add_rule("/", self.echo_backend("root"))
        ingress.add_rule("/foo", self.echo_backend("foo"))
        ingress.add_rule("/foo/bar", self.echo_backend("foo-bar"))

        ingress.add_host_default_backend("my.host", self.echo_backend("my.host/hey"))

    def echo_backend(self, text):
        deploy = kplus.Deployment(
            self,
            text,
            containers=[
                kplus.ContainerProps(
                    image="hashicorp/http-echo",
                    args=["-text", text],
                    port_number=5678,
                )
            ],
        )

        return kplus.IngressBackend.from_service(
            deploy.expose_via_service(ports=[kplus.ServicePort(port=5678)])
        )


app = App()
HttpEcho(app, "http-echo")

app.synth()
